File: java/89.gray-code.java
/*
 * @lc app=leetcode id=89 lang=java
 *
 * [89] Gray Code
 */

// @lc code=start

import java.util.ArrayList;

class Solution {
    public List<Integer> grayCode(int n) {
        ArrayList<Integer> arr = new ArrayList<Integer>();
        arr.add(0);
        arr.add(1);
        
        for(int i=1; i<n; i++) {
            int sum = (int) Math.pow(2, i);

            for (int j=arr.size()-1;j>=0;j--) {
                arr.add(arr.get(j) + sum);
            }
        }

        return arr;
    }
}
// @lc code=end

File: python3/89.gray-code.py
#
# @lc app=leetcode id=89 lang=python3
#
# [89] Gray Code
#

# @lc code=start
class Solution:
    def grayCode(self, n: int) -> List[int]:
        start_list = [0, 1]
        for i in range(1, n):
            summand = 2**(i)
            start_list = start_list+[elem+summand for elem in reversed(start_list)]
        return start_list

        
# @lc code=end
